Issue 1045 - Allow downcasting of parquet timestamp[ns] to timestamp[us] for Table.add_files()

lloyd-EA opened the pull request:
When calling Table.add_files(), the env PYICEBERG_DOWNCAST_NS_TIMESTAMP_TO_US_ON_WRITE isn't being used on _check_pyarrow_schema_compatible, therefore, it fails for parquet files with ns timestamps.

Credits to @fussion2222 for suggesting this fix.

kevinjqliu reviewed (commented):
Thanks for the PR!

Generally LGTM, another option is to pass the option down from `add_files`. 

Could you also add a test case for this? 
Something similar to,
https://github.com/apache/iceberg-python/blob/3b53edc2ec523782735e8ddf54119ab8ef3c953d/tests/integration/test_add_files.py#L588

lloyd-EA commented:
@kevinjqliu Thanks for having a look.

I've updated the PR and added a test case for it as requested. 

I'm open for suggestions and corrections. Thanks!

sungwy commented:
Thanks for working on this PR @lloyd-EA and thank you @kevinjqliu for the review. Could I add another suggestion to read the table using Spark and verify that the value that's been read is equal to the expected value?

As I mentioned in a comment in https://github.com/apache/iceberg-python/issues/1045#issuecomment-2617788267 I think it would be great to verify that the data is readable as `us` timestamp through a Spark session as well. Here's an example that uses a spark session for reference: https://github.com/apache/iceberg-python/blob/985029042199d870f25b6fbec0e80907d4440f41/tests/integration/test_add_files.py#L183

lloyd-EA commented:
@sungwy I updated the PR there. It seems that Spark is reads in the column as `us` and doesn't check the underlying data file's precision being `ns`, so when it is converted to a `datetime` the value is wrong.

expected: `2021-03-17 07:54:47.249846175`
![image](https://github.com/user-attachments/assets/9e6056b9-ce1e-4e89-a16e-c48bde478934)

To me, this is a problem with Spark not being able to handle `ns` timestamps properly, and since I won't be using Spark for my application, I'll continue to use my forked repo for now as it's sufficient for my application.

sungwy commented:
Thank you for following up on the suggestion @lloyd-EA . I'm glad we included this integration test to find out the issue with Spark's timestamp conversion.

I think we will have to investigate the corresponding schema evolution handling logic in Spark-Iceberg to understand if this is a bug, or an expected behavior with Spark.

If it is not possible to fix this issue in Spark-Iceberg, I think it'll mean that we will not be able to support adding `ns` timestamp files until [V3 spec](https://iceberg.apache.org/spec/#version-3-extended-types-and-capabilities) is supported.

Please do keep in mind that Iceberg spec is a language agnostic spec, and hence ruling out the possibility of using the same Iceberg table with other tools (like Trino or Spark) may impact you down the line when your application system grows.

lloyd-EA commented:
Understood, thanks!
